File: my-solutions/0074-search-a-2d-matrix/solution.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:

        def binSearch(row):
            l, r = 0, len(matrix[0]) - 1
            # print(matrix[row])
            while(l <= r):
                mid = int((l+r)/2)
                if matrix[row][mid] == target:
                    return mid
                elif matrix[row][mid] > target:
                    r = mid - 1
                else:
                    l = mid + 1
            return -1

        l, r = 0, len(matrix) - 1
        while(l <= r):
            mid = int((l + r) / 2)
            # print(l, mid, r)
            if target <= matrix[mid][-1] and target >= matrix[mid][0]:
                return binSearch(mid) != -1
            elif target > matrix[mid][-1]:
                l = mid + 1
            else:
                r = mid - 1
        
        return False
        

File: my-solutions/0074-search-a-2d-matrix/solution.cpp
class Solution {
public:
    bool searchArray(vector<vector<int>>& matrix, int row, int target) {
        // cout << row << " " << target << endl;
        int l = 0, r = size(matrix[row]) - 1;
        // cout << l << " " << r << endl;
        while(l <= r) {
            int mid = (l + r) / 2;
            // cout << l << " " << mid << " " << r << endl;
            // cout << target << " " << matrix[row][mid] << endl;
            if(target == matrix[row][mid])
                return true;
            else if(target > matrix[row][mid])
                l = mid + 1;
            else
                r = mid - 1;
        }
        return false;
    }
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if(size(matrix) == 1 && size(matrix[0]) == 1)
            return matrix[0][0] == target;
        if (size(matrix) == 1)
            return searchArray(matrix, 0, target);
        int l = 1, r = size(matrix) - 1;
        while(l <= r) {
            int row = (l + r) / 2;
            if(target == matrix[row][0] || target == matrix[row-1][0])
                return true;
            else if(target > matrix[row-1][0] && target < matrix[row][0])
                return searchArray(matrix, row - 1, target);
            else if(target > matrix[row][0])
                l = row + 1;
            else
                r = row - 1;
        }

        return searchArray(matrix, size(matrix) - 1, target);
    }
};
